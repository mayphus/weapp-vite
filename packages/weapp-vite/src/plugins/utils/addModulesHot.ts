import type { PluginContext } from 'rollup'

export function addModulesHot(entriesSet: Set<string>, pluginContext: PluginContext) {
  for (const entry of entriesSet) {
    const moduleInfo = pluginContext.getModuleInfo(entry)

    if (moduleInfo) {
      const stack = [moduleInfo.id] // 用栈模拟递归
      const visitedModules = new Set<string>()

      while (stack.length > 0) {
        const id = stack.pop()

        if (id && !visitedModules.has(id)) {
          visitedModules.add(id)

          const info = pluginContext.getModuleInfo(id)

          if (info) {
            pluginContext.addWatchFile(info.id)
            // 将子依赖加入栈
            stack.push(...info.importedIds)
          }
        }
      }
    }
  }
}
