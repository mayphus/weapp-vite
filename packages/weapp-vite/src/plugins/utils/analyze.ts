import type { App, Component, Page } from '@weapp-core/schematics'
import { get } from '@weapp-core/shared'

export function analyzeAppJson(json: App) {
  const entries: string[] = []
  const pages = json.pages ?? []
  const components = Object.values(json.usingComponents ?? {})
  const subPackages = (
    [...json.subPackages ?? [], ...json.subpackages ?? []]?.reduce<string[]>(
      (acc, cur) => {
        acc.push(...(cur.pages ?? []).map(x => `${cur.root}/${x}`))
        return acc
      },
      [],
    )) ?? []

  if (get(json, 'tabBar.custom')) {
    entries.push('custom-tab-bar/index')
  }
  // 全局工具栏
  // https://developers.weixin.qq.com/miniprogram/dev/framework/runtime/skyline/appbar.html
  if (get(json, 'appBar')) {
    entries.push('app-bar/index')
  }
  entries.push(
    ...pages,
    ...components,
    ...subPackages,
  )
  return entries
}

export function analyzeCommonJson(json: Page | Component) {
  const entries: string[] = []
  const components = Object.values(json.usingComponents ?? {})
  entries.push(...components)
  return entries
}
