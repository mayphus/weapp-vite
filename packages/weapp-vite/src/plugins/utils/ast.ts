import type { Node } from 'estree'
// import type { ProgramNode } from 'rollup'
import { walk } from 'estree-walker'

export function collectRequireTokens(ast: Node) {
  const requireModules: {
    start: number
    end: number
    value: string
    async?: boolean
    // leadingComment: string
  }[] = []
  walk(ast, {
    enter(node) {
      if (node.type === 'CallExpression') {
        // TODO 可能要被废弃
        // https://developers.weixin.qq.com/miniprogram/dev/reference/api/require.html#require-async-%E9%93%BE%E5%BC%8F%E8%B0%83%E7%94%A8
        if (node.callee.type === 'Identifier' && node.callee.name === 'require') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
            })
          }
        }
        else if (
          node.callee.type === 'MemberExpression'
          && node.callee.object.type === 'Identifier'
          && node.callee.object.name === 'require'
          && node.callee.property.type === 'Identifier'
          && node.callee.property.name === 'async') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
              async: true,
            })
          }
        }
      }
    },
  })

  return {
    requireModules,
  }
}
