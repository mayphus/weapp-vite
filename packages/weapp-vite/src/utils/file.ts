import { configExtensions, jsExtensions, templateExtensions, vueExtensions } from '@/constants'
import fs from 'fs-extra'
import path from 'pathe'

export function isJsOrTs(name?: string) {
  if (typeof name === 'string') {
    return jsExtensions.some(x => name.endsWith(`.${x}`))
  }
  return false
}

export function isTemplateRequest(request: string) {
  return request.endsWith('.wxml') || request.endsWith('.html')
}

export function changeFileExtension(filePath: string, extension: string) {
  if (typeof filePath !== 'string') {
    throw new TypeError(`Expected \`filePath\` to be a string, got \`${typeof filePath}\`.`)
  }

  if (typeof extension !== 'string') {
    throw new TypeError(`Expected \`extension\` to be a string, got \`${typeof extension}\`.`)
  }

  if (filePath === '') {
    return ''
  }

  extension = extension ? (extension.startsWith('.') ? extension : `.${extension}`) : ''

  const basename = path.basename(filePath, path.extname(filePath))
  return path.join(path.dirname(filePath), basename + extension)
}

export async function findVueEntry(filepath: string) {
  for (const ext of vueExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsEntry(filepath: string) {
  for (const ext of jsExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsonEntry(filepath: string) {
  for (const ext of configExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findTemplateEntry(filepath: string) {
  for (const ext of templateExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export function isTemplate(filepath: string) {
  return templateExtensions.some(ext => filepath.endsWith(`.${ext}`))
}

export function touch(filename: string) {
  const time = new Date()

  try {
    fs.utimesSync(filename, time, time)
  }
  catch {
    fs.closeSync(fs.openSync(filename, 'w'))
  }
}
