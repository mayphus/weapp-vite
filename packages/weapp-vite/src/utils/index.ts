export * from './file'
export * from './json'
export * from './projectConfig'
export * from './scan'
// export * from './wxml'
// export * from './wxs'
