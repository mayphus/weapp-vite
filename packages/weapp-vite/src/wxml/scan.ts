import type { ComponentsMap, ScanWxmlOptions, WxmlDep } from '@/types'
import type { Buffer } from 'node:buffer'
import type { Token } from './shared'
import { isBuiltinComponent } from '@/auto-import-components/builtin'
import { jsExtensions } from '@/constants'
import { defu, objectHash } from '@weapp-core/shared'
import { Parser } from 'htmlparser2'
import { LRUCache } from 'lru-cache'
import { srcImportTagsMap } from './shared'

export interface WxmlToken {
  components: ComponentsMap
  deps: WxmlDep[]
  removeStartStack: number[]
  removeEndStack: number[]
  commentTokens: Token[]
  inlineWxsTokens: Token[]
  wxsImportNormalizeTokens: Token[]
  removeWxsLangAttrTokens: Token[]
  eventTokens: Token[]
  code: string
}

export const scanWxmlCache = new LRUCache<string, WxmlToken>(
  {
    max: 512,
  },
)

function getCacheKey(wxml: string | Buffer, options?: ScanWxmlOptions) {
  return objectHash({
    wxml: wxml.toString(),
    options,
  })
}

export function defaultExcludeComponent(tagName: string) {
  return isBuiltinComponent(tagName)
}

export function scanWxml(wxml: string | Buffer, options?: ScanWxmlOptions) {
  const opts = defu<Required<ScanWxmlOptions>, ScanWxmlOptions[]>(options, {
    excludeComponent: defaultExcludeComponent,
    platform: 'weapp',
  })
  const cacheKey = getCacheKey(wxml, opts)
  const t = scanWxmlCache.get(cacheKey)
  if (t) {
    return t
  }
  const ms = wxml.toString()
  const deps: WxmlDep[] = []
  let currentTagName: string | undefined
  let importAttrs: undefined | string[]
  let attrs: Record<string, string> = {}
  const components: ComponentsMap = {}
  let tagStartIndex = 0
  // transformOn
  // https://github.com/vuejs/core/blob/76c43c6040518c93b41f60a28b224f967c007fdf/packages/compiler-core/src/transforms/vOn.ts

  // 条件编译注释
  const removeStartStack: number[] = []
  const removeEndStack: number[] = []
  // 注释
  const commentTokens: Token[] = []
  // 内联wxs
  const inlineWxsTokens: Token[] = []
  // 处理 xxx.wxs.ts 转变为合法引用
  const wxsImportNormalizeTokens: Token[] = []
  // 移除内联 wxs 的 lang 属性
  const removeWxsLangAttrTokens: Token[] = []
  // 事件转义
  const eventTokens: Token[] = []
  // tag 调用栈
  const tagStack: string[] = []
  const parser = new Parser(
    {
      onopentagname(name) {
        tagStack.push(name)
        currentTagName = name
        importAttrs = srcImportTagsMap[currentTagName]
        tagStartIndex = parser.startIndex
      },
      onattribute(name, value, quote) {
        attrs[name] = value
        if (importAttrs && currentTagName) {
          for (const attrName of importAttrs) {
            if (attrName === name) {
              deps.push({
                name,
                value,
                quote,
                tagName: currentTagName,
                start: parser.startIndex,
                end: parser.endIndex,
                attrs,
              })
              if (currentTagName === 'wxs' && name === 'src') {
                if (/\.wxs.[jt]s$/.test(value)) {
                  // 5 是 'src="'.length
                  // 1 是 '"'.length
                  wxsImportNormalizeTokens.push(
                    {
                      start: parser.startIndex + 5,
                      end: parser.endIndex - 1,
                      value,
                    },
                  )
                }
              }
            }
          }
        }
        // 事件绑定
        if (name.startsWith('@')) {
          const start = parser.startIndex
          const end = parser.startIndex + name.length
          const { dir, mods } = name.split('.')
            .filter(x => x)
            .reduce<{ dir: string, mods: Record<string, boolean> }>((acc, cur, idx) => {
              if (idx === 0) {
                acc.dir = cur
              }
              else {
                acc.mods[cur] = true
              }
              return acc
            }, { dir: '', mods: {} })

          let rep: string
          if (mods.catch && mods.capture) {
            rep = `capture-catch:${dir.slice(1)}`
          }
          else if (mods.catch) {
            rep = `catch:${dir.slice(1)}`
          }
          else if (mods.mut) {
            rep = `mut-bind:${dir.slice(1)}`
          }
          else if (mods.capture) {
            rep = `capture-bind:${dir.slice(1)}`
          }
          else {
            rep = `bind:${dir.slice(1)}`
          }
          if (rep) {
            eventTokens.push({
              start,
              end,
              value: rep,
            })
          }
        }
        // 移除内联 wxs 的 lang
        if (currentTagName === 'wxs' && name === 'lang' && jsExtensions.includes(value)) {
          removeWxsLangAttrTokens.push({
            start: parser.startIndex,
            end: parser.endIndex,
            value,
          })
        }
      },
      onclosetag() {
        currentTagName = tagStack.pop()
        if (currentTagName && !opts.excludeComponent(currentTagName)) {
          if (Array.isArray(components[currentTagName])) {
            components[currentTagName].push({
              start: tagStartIndex,
              end: parser.endIndex + 1,
            })
          }
          else {
            components[currentTagName] = [{
              start: tagStartIndex,
              end: parser.endIndex + 1,
            }]
          }
        }

        currentTagName = ''
        attrs = {}
        importAttrs = undefined
        tagStartIndex = 0
      },
      ontext(data) {
        if (currentTagName === 'wxs' && jsExtensions.includes(attrs.lang)) {
          inlineWxsTokens.push({
            start: parser.startIndex,
            end: parser.endIndex,
            value: data,
          })
        }
      },
      // <!--  #ifdef  %PLATFORM% -->
      // 平台特有的组件
      // <!--  #endif -->
      oncomment(data) {
        let match = /#ifdef\s+(\w+)/.exec(data)
        if (match) {
          if (match[1] !== opts.platform) {
            removeStartStack.push(parser.startIndex)
          }
        }
        match = /#endif/.exec(data)
        if (match) {
          removeEndStack.push(parser.endIndex + 1)
        }
        commentTokens.push({
          start: parser.startIndex,
          end: parser.endIndex + 1,
          value: data,
        })
      },
    },
    {
      lowerCaseTags: false,
      xmlMode: true,
    },
  )
  parser.write(
    ms,
  )
  parser.end()

  const token: WxmlToken = {
    components,
    deps,
    removeStartStack,
    removeEndStack,
    commentTokens,
    inlineWxsTokens,
    wxsImportNormalizeTokens,
    removeWxsLangAttrTokens,
    eventTokens,
    code: ms,
  }
  scanWxmlCache.set(cacheKey, token)
  return token
}

export type ScanWxmlResult = ReturnType<typeof scanWxml>
