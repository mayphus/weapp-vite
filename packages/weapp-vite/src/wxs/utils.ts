import { addExtension, removeExtensionDeep } from '@weapp-core/shared'

export function normalizeWxsFilename(value: string) {
  return addExtension(removeExtensionDeep(value), '.wxs')
}
