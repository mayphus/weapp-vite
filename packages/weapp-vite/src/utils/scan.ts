import fs from 'fs-extra'
import { jsExtensions } from '../constants'
import { changeFileExtension } from '../utils'

// https://developers.weixin.qq.com/miniprogram/dev/framework/structure.html
// app: js + json
export async function searchPageEntry(wxmlPath: string) {
  if (await fs.exists(wxmlPath)) {
    for (const ext of jsExtensions) {
      const entryPath = changeFileExtension(wxmlPath, ext)
      if (await fs.exists(entryPath)) {
        return entryPath
      }
    }
  }
}
// https://developers.weixin.qq.com/miniprogram/dev/framework/structure.html
// page: wxml + js
export async function isPage(wxmlPath: string) {
  return Boolean(await searchPageEntry(wxmlPath))
}

export interface SearchAppEntryOptions {
  root: string
  formatPath?: (p: string) => string
}

// component: wxml + js + json + json.component === true
export async function isComponent(wxmlPath: string) {
  if (await isPage(wxmlPath)) {
    const jsonPath = changeFileExtension(wxmlPath, 'json')
    if (await fs.exists(jsonPath)) {
      const json = await fs.readJson(jsonPath, { throws: false })
      if (json && json.component) {
        return true
      }
    }
  }
  return false
}
