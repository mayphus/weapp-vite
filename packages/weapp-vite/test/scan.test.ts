import { isComponent } from '@/utils/scan'
import { fdir as Fdir } from 'fdir'
import path from 'pathe'
import { appsDir } from './utils'

describe('scan', () => {
  describe.skip('isComponent', () => {
    it('vite-native components', async () => {
      const cwd = path.resolve(appsDir, 'vite-native')
      const dir = path.resolve(cwd, 'components')
      const relFiles = await new Fdir()
        .withRelativePaths()
        .globWithOptions(
          ['**/*.{wxml,html}'],
          {
            cwd: dir,
            ignore: [],
          },
        )
        .crawl(dir)
        .withPromise()
      expect(relFiles.sort()).toMatchSnapshot()
      expect(
        (await Promise.all(relFiles.map(async (x) => {
          return {
            isComponent: await isComponent(path.join(dir, x)),
            result: x,
          }
        }))).filter(x => x.isComponent).map(x => x.result),
      ).toMatchSnapshot()
    })
  })
})
