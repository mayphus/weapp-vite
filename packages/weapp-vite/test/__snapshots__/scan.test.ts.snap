// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scan > isComponent > vite-native components 1`] = `
[
  "HelloWorld/HelloWorld.wxml",
  "HelloWorld/index.wxml",
  "HiChina/HiChina.wxml",
  "avatar/avatar.wxml",
  "icebreaker/index.wxml",
  "navigation-bar/navigation-bar.wxml",
  "test.wxml",
]
`;

exports[`scan > isComponent > vite-native components 2`] = `
[
  "HelloWorld/HelloWorld.wxml",
  "HelloWorld/index.wxml",
  "HiChina/HiChina.wxml",
  "icebreaker/index.wxml",
]
`;
